### PR TITLE
Feature/backend/employee service

### LINKED ISSUE #13: Employees cannot be updated properly to a service
Updating employee information to a service does not remove old information present in the database.

### PR DESCRIPTION
**Pull request by Alex-DSJ**
# Description

create/modify services based on different durations and prices
implement basic CRUD operations for EMPMAN-1 module

---

# Related Jira Issue (ID/URL)

[RTTER-299 BE - Add/Remove employee from existing service and duration/price changes](https://rtterror.atlassian.net/jira/software/projects/RTTER/boards/1?selectedIssue=RTTER-299)
[RTTER-192 BE - ROUTE for CRUD operation on employees’ basic information](https://rtterror.atlassian.net/jira/software/projects/RTTER/boards/1?selectedIssue=RTTER-192)
[RTTER-193 BE - SCHEMA for employees’ basic information](https://rtterror.atlassian.net/jira/software/projects/RTTER/boards/1?selectedIssue=RTTER-193)
[RTTER-194 BE - DAO for CRUD operation on employees’ basic information](https://rtterror.atlassian.net/jira/software/projects/RTTER/boards/1?selectedIssue=RTTER-194)
[RTTER-195 BE - SERVICE for CRUD operation on employees’ basic information](https://rtterror.atlassian.net/jira/software/projects/RTTER/boards/1?selectedIssue=RTTER-195)
[RTTER-196 DB - Create a table in MySQL to store employees’ basic information](https://rtterror.atlassian.net/jira/software/projects/RTTER/boards/1?selectedIssue=RTTER-196)

# Related GitHub Issues

- This fixes #13.

# Checklist

- [X] Code compiles correctly
- [ ] Code is covered by current tests or new tests
- [X] Code is formatted correctly
- [ ] Extended the README / documentation, if necessary

---

# Comment/Discussion

NA

### CODE REVIEW

**Comment by NJLprograms** [comment]
E2E checks and SonarCloud Quality Gate have failed, you can view the Checks tab to see why the sonarcloud analysis failed

File: backend/src/interfaces/IService.ts [hunk 1]
```diff
@@ -25,4 +26,25 @@ export interface IServiceDto {
...
+  employee_ids?: string[];
 }
+
+interface IDurationsPrices {
+  id?: string;    //  for updating a series of services
+  duration: number;
+  price: string;
+}
+
+export interface IServicesDto {
```
Comment (yuxiang-zhang) [inline backend/src/interfaces/IService.ts:38]: What are some of the use cases for updating multiple services at once? This interface looks like a repeat of `IServiceDto` and is a burden to maintain.

File: backend/src/routes/employee.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+});
+
+router.post('/', (req, res) => {
+    const { body } = req;
+    employeeService.createItem(body)
+        .then((newEmployee) => {
+            res.status(200).send(newEmployee);
+        })
+        .catch((error) => {
+            res.status(400).send(error);
```
Comment (yuxiang-zhang) [inline backend/src/routes/employee.ts:18]: Referring to #29, this should probably follow the same adjustment.

Comment (yuxiang-zhang) [reply backend/src/routes/employee.ts:18]: similarly for L. 28, 39, 48, 58

File: backend/src/services/combo-service.ts [hunk 3]
```diff
@@ -142,6 +142,32 @@ class ComboService extends GeneralService<ICombo, IComboDto> {
...
+      await this.model.destroy({
+        where: {
+          id: comboId
+        },
+        transaction: t
+      });
+      counter++;
+    }
+    console.log(counter + 'combos has been deleted')
+    let leftComboIds = await this.serviceComboModel.findAll({
```
Comment (yuxiang-zhang) [inline backend/src/services/combo-service.ts:164]: It'd be good practice to use `const` whenever the variable doesn't change their value.

**Review by yuxiang-zhang — COMMENTED** [review]
I see that this is still work in progress. More review is needed after the entirety is synced on the remote branch.

File: backend/src/interfaces/IService.ts (hunk 1 above)
Comment (Alex-DSJ) [reply backend/src/interfaces/IService.ts:38]: this interface is to define necessary fields for creating a service, any service info unrelated with the id should satisfy this interface

File: backend/src/routes/employee.ts (hunk 2 above)
Comment (Alex-DSJ) [reply backend/src/routes/employee.ts:18]: sure, will modify this

File: backend/src/services/combo-service.ts (hunk 3 above)
Comment (Alex-DSJ) [reply backend/src/services/combo-service.ts:164]: agree, thanks for you reviewing

File: backend/src/interfaces/IService.ts (hunk 1 above)
Comment (yuxiang-zhang) [reply backend/src/interfaces/IService.ts:38]: What is the difference between `IServicesDto` and `IServiceDto`?

File: backend/src/models/employee.ts [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        allowNull: false
+    },
+    gender: {
+        type: DataTypes.STRING,
+    },
+    dob: {
+        type: DataTypes.STRING,
+        allowNull: false
+    },
+    sin: {
```
Comment (kimchhengheng) [inline backend/src/models/employee.ts:27]: Should we stored Sin in our DB?

Comment (Alex-DSJ) [reply backend/src/models/employee.ts:27]: I didn't find any regulation regarding to the storage of SIN, so I think it depends on the requirements of the stakeholder?

**Review by kimchhengheng — APPROVED** [review]

**Comment by Alex-DSJ** [comment]
> E2E checks and SonarCloud Quality Gate have failed, you can view the Checks tab to see why the sonarcloud analysis failed

thanks for reminding me, I refactored the route by extracting the error process method, and remove the useless method due to the change of requirement.

**Review by kevinlizh1992 — APPROVED** [review]

**Review by yuxiang-zhang — COMMENTED** [review]
These changes should solve #39. Build is currently failing because the test coverage doesn't meet the target threshold of 80%. It is recommended to merge as tests are assigned to a different developer and will be added in a separate PR

File: backend/src/interfaces/IService.ts (hunk 1 above)
Comment (Alex-DSJ) [reply backend/src/interfaces/IService.ts:38]: `IServiceDto` has been changed into `IServicesDto`, the later one contains all neccessary information to create a group of services.

**Comment by yuxiang-zhang** [comment]
Merged as per previous reviews and new PR #41 merged

**Pull request merged**